### PR TITLE
Objectness IoU Sort

### LINKED ISSUE #3605: I found Objectness in ComputeLoss which might need improvement
Hi,
sorry for disturbing you. I found one part in your code which might be improvement.
There maybe several GTs match the same anchor when calculate `ComputeLoss` in the scene with dense targets.

[`# Objectness`<br/>`tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio`](https://github.com/ultralytics/yolov5/blob/31336075609a3fbcb4afe398eba2967b22056bfa/utils/loss.py#L136)

The `tobj` is filled with the location of `[b, a, gj, gi]` which will be covered with the next same location. 
The final tobj is the bottom right GT in `build_targets`.
It may be not reasonable. The improvement of matching the GT with largest IOU could be used with the following modification.

source:
`tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio`
should be replaced with:
```
score_iou = iou.detach().clamp(0).type(tobj.dtype)
sort_id = torch.argsort(score_iou)
b, a, gj, gi, score_iou = b[sort_id], a[sort_id], gj[sort_id], gi[sort_id], score_iou[sort_id]
tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * score_iou  # iou ratio
```

I just tried in the private dataset with the scene with very dense objects, the performance will be improved with above change.

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved objectness target calculation in loss function for better training stability.

### 📊 Key Changes
- Additional sorting step before calculating the objectness target values.
- Sorting IOU scores and the corresponding indices before assigning them to `tobj`.

### 🎯 Purpose & Impact
- **Purpose**: To enhance the assignment of objectness scores during model training, by ensuring the targets are set in a sorted order based on IOU scores.
- **Impact**: This could potentially lead to more stable and consistent training results, as the objectness targets would be less noisy and more reflective of the true IOU values.

👩‍💻🧑‍💻 Users can expect smoother training convergence and possibly improved object detection performance from their YOLOv5 models. 🎯🚀

### CODE REVIEW

**Review by XHBrain — APPROVED** [review]

**Comment by glenn-jocher** [comment]
@XHBrain I've got an update here. I've got a YOLOv5l on COCO test run going on a 4x V100 instance here:
https://wandb.ai/glenn-jocher/study-mixup-coco

So far the IoU sort seems to have minimal effect, but we'll need to wait several more days for full results.
Screenshot 2021-06-17 at 21 20 19

**Comment by glenn-jocher** [comment]
@XHBrain PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by glenn-jocher** [comment]
Followed up with #3786
